File: julia_fractal/julia_to_pict.cpp
#include <iostream>
#include <cstring>
#include <cstdint>
#include <cmath>

struct dvec4 { double x=0.0,y=0.0,z=0.0,w=0.0; };
struct dvec2{double x=0.0,y=0.0; dvec2(const double& vx, const double& vy) : x(vx),y(vy){}};

#define dmin(a, b) (((a)<(b)) ? (a) : (b))
#define dmax(a, b) (((a)>(b)) ? (a) : (b))
#define dclamp(v, inf, sup) dmin(dmax(v,inf),sup)

dvec4 ComputeImage(const dvec2& vFragCoord, const dvec2& vSize)
{
	dvec4 pixelColor;
	const double& zoom = 0.93;
	const double& threshold = 0.14;
	const double mns = zoom / dmin(vSize.x, vSize.y);
	double zx = (vFragCoord.x * 2.0 - (double)vSize.x) * mns;
	double zy = (vFragCoord.y * 2.0 - (double)vSize.y) * mns;
	const double& px = 1.0;
	const double& py = 0.3;
	double d = 0.0;
	for (int i = 0; i < 12; i++)
	{
		d = zx * zx + zy * zy;
		if (d < 4.0)                        // 4 mean 2 * 2 (opt fix with dot instead of length
		{
			double tx = zx;                 // tmp x
			zx = zx * zx - zy * zy - px;    // julia zx
			zy = 2.0 * tx * zy - py;        // julia zy
		}
	}
	
	// distance field formula
	const double de = sqrt(d) * log(d);
	const double dd = 1.0 / d;
	
	pixelColor.x = dclamp(0.15 * dd + de, 0.0, 1.0);
	pixelColor.y = dclamp(0.1 * dd + de, 0.0, 1.0);
	pixelColor.z = dclamp(0.1 * dd + de, 0.0, 1.0);
	pixelColor.w = dclamp(1.0 * dd + de, 0.0, 1.0);
	return pixelColor;
}

int main ()
{
	const int& w = 2560;
	const int& h = 1440;
	
	const size_t datasize = 3*w*h;
	auto img = new uint8_t[datasize];
	memset(img,0,datasize);
	
	size_t x,y;
	dvec4 pixelColor;
	for(int i=0; i<w; ++i) {
		for(int j=0; j<h; ++j) {
			x=i; y=(h-1)-j;
			pixelColor = ComputeImage(dvec2((double)x,(double)y), dvec2((double)w,(double)h));
			img[(x+y*w)*3+2] = (uint8_t)(pixelColor.x * 255.0);
			img[(x+y*w)*3+1] = (uint8_t)(pixelColor.y * 255.0);
			img[(x+y*w)*3+0] = (uint8_t)(pixelColor.z * 255.0);
		}
	}
	
	const size_t headersize = 54;
	const size_t filesize = headersize + datasize;
	
	unsigned char bmpfileheader[14] = {'B','M', 0,0,0,0, 0,0, 0,0, headersize,0,0,0};
	bmpfileheader[2] = (uint8_t)(filesize);
	bmpfileheader[3] = (uint8_t)(filesize>> 8);
	bmpfileheader[4] = (uint8_t)(filesize>>16);
	bmpfileheader[5] = (uint8_t)(filesize>>24);
	
	unsigned char bmpinfoheader[40] = {40,0,0,0, 0,0,0,0, 0,0,0,0, 1,0, 24,0};
	bmpinfoheader[4] = (uint8_t)(w);
	bmpinfoheader[5] = (uint8_t)(w>> 8);
	bmpinfoheader[6] = (uint8_t)(w>>16);
	bmpinfoheader[7] = (uint8_t)(w>>24);
	bmpinfoheader[8] = (uint8_t)(h);
	bmpinfoheader[9] = (uint8_t)(h>> 8);
	bmpinfoheader[10] = (uint8_t)(h>>16);
	bmpinfoheader[11] = (uint8_t)(h>>24);
	
	unsigned char bmppad[3] = {0,0,0};
	auto f = fopen("img.bmp","wb");
	fwrite(bmpfileheader,1,14,f);
	fwrite(bmpinfoheader,1,40,f);
	for(int i=0; i<h; ++i)
	{
		fwrite(img+(w*(h-i-1)*3),3,w,f);
		fwrite(bmppad,1,(4-(w*3)%4)%4,f);
	}
	fclose(f);
	
	return 0;
}

File: julia_to_pict.rs
use std::f64;
use std::vec;
use std::fs::File;
use std::io::Write;

struct Dvec3 { x: f64, y: f64, z: f64 }
impl Dvec3 {
    fn new(vx: f64, vy: f64, vz: f64) -> Dvec3 { Dvec3 { x: vx, y: vy, z: vz } }
}

fn compute_image(cx:f64, cy:f64, sx:f64, sy:f64) -> Dvec3
{
    let zoom = 0.93;
    let mns = zoom / f64::min(sx, sy);
    let mut zx = (cx * 2.0 - sx) * mns;
    let mut zy = (cy * 2.0 - sy) * mns;
    let px = 1.0;
    let py = 0.3;

    let mut d = 0.0;
    let mut tmp_x;
    for _n in 1..12
    {
        d = zx * zx + zy * zy;
        if d < 4.0
        {
            tmp_x = zx;
            zx = zx * zx - zy * zy - px;
            zy = 2.0 * tmp_x * zy - py;
        }
    }

    let de = f64::sqrt(d) * f64::log(d, 10.0);
    let dd = 1.0 / d;

    return Dvec3::new(
        f64::clamp(0.15 * dd + de, 0.0, 1.0),
        f64::clamp(0.1 * dd + de, 0.0, 1.0),
        f64::clamp(0.1 * dd + de, 0.0, 1.0),
    );
}

fn generate_shader(w:usize, h:usize) -> Vec<u8>
{
    let datasize = 3 * w * h;
    let mut final_img:Vec<u8> = vec![0; datasize];
    for x in 0..w {
        for j in 0..h {
            let y = (h-1)-j;
            let pixel_color = compute_image(x as f64, y as f64, w as f64, h as f64);
            final_img[(x+y*w)*3+2] = (pixel_color.x * 255.0) as u8;
            final_img[(x+y*w)*3+1] = (pixel_color.y * 255.0) as u8;
            final_img[(x+y*w)*3+0] = (pixel_color.z * 255.0) as u8;
        }
    }

    final_img
}

fn generate_bitmap(w:usize, h:usize, img:Vec<u8>) -> std::io::Result<()>
{
    let headersize = 54;
    let filesize = headersize as usize + img.len();

    let mut bmp_file_header:Vec<u8> = vec![0; 14];
    bmp_file_header[0] = 'B' as u8;
    bmp_file_header[1] = 'M' as u8;
    bmp_file_header[2] = (filesize) as u8;
    bmp_file_header[3] = (filesize >> 08) as u8;
    bmp_file_header[4] = (filesize >> 16) as u8;
    bmp_file_header[5] = (filesize >> 24) as u8;
    bmp_file_header[10] = headersize;

    let mut bmp_info_header:Vec<u8> = vec![0; 40];
    bmp_info_header[0] = 40;
    bmp_info_header[4] = (w) as u8;
    bmp_info_header[5] = (w>>8) as u8;
    bmp_info_header[6] = (w>>16) as u8;
    bmp_info_header[7] = (w>>24) as u8;
    bmp_info_header[8] = (h) as u8;
    bmp_info_header[9] = (h>>8) as u8;
    bmp_info_header[10] = (h>>16) as u8;
    bmp_info_header[11] = (h>>24) as u8;
    bmp_info_header[12] = 1;
    bmp_info_header[14] = 24;

    let bmp_pad:Vec<u8> = vec![0; 3];

    let mut file = File::create("julia.bmp")?;
    file.write_all (&bmp_file_header)?;
    file.write_all(&bmp_info_header)?;

    for i in 0..h {
        let pos = w*(h-i-1)*3;
        file.write(&img[pos..(pos+3*w)])?;
        file.write(&bmp_pad[..((4 - (w * 3) % 4) % 4)])?;
    }
    file.sync_all()?;
    Ok(())
}

fn main() -> std::io::Result<()>
{
    let w:usize = 2560;
	let h:usize = 1440;

	generate_bitmap(w,h, generate_shader(w,h))?;

    Ok(())
}
